### ISSUE TITLE
Design new Python API to replace Legacy API

### ISSUE THREAD

**Issue by asottile** [PyCQA/flake8#140]
In GitLab by @sigmavirus24 on Aug 3, 2016, 06:13

# Description

Now that Flake8 3.0 is released and our old API is considered "legacy" (and deprecated of sorts), let's work together to design a Python API that is powerful and elegant and extensible.

# Use Cases

- Need a way to pass the lines of a file into Flake8 (see https://gitlab.com/pycqa/flake8/issues/208#note_14544578)
- Need a way to iterate over the errors collected by Flake8
- Need objects to be easily serializable

**Comment by asottile** [comment]
In GitLab by @sigmavirus24 on Aug 3, 2016, 06:31

Some feedback from [twitter](https://twitter.com/sigmavirus24/status/760828828029558785):

```
tl;dr: I want to pass in a string containing code and get a data structure enumerating the feedback.
```

```
feedback should be an easy to JSONify object so it's easy to stick behind a web API somewhere if need be. ;-)
```

**Comment by asottile** [comment]
In GitLab by @sigmavirus24 on Aug 10, 2016, 17:35

@xZise what about adding an equivalent to `stdin_get_value` to the new public API for Flake8?

**Comment by asottile** [comment]
In GitLab by @asottile on Aug 24, 2016, 16:23

I'm saddened and frustrated that the old api was removed without any working replacement :(

**Comment by asottile** [comment]
In GitLab by @sigmavirus24 on Aug 27, 2016, 10:07

@asottile thank you for expressing your frustration. Given cheetah-lint has the most complex of use-cases I've seen of anyone using, would you provide some use cases for cheetah lint? What do you need to do with flake8?

From your myriad of other issues it seems you need a way to provide lines to be checked. What else?

**Comment by asottile** [comment]
In GitLab by @sigmavirus24 on Aug 27, 2016, 10:08

For what it's worth, I didn't want to redesign a new API because I have zero use for one. I wanted users to help provide the use cases for the API so users could have a say in what the stable API was. I also don't want users forcing us into API by importing things like was done with `flake8.engine`.

**Comment by asottile** [comment]
In GitLab by @asottile on Aug 27, 2016, 10:42

I think the only thing I need from flake8 is a way to pass in source lines and receive a datastructure containing error codes.  I was (admittedly) hackily doing this in 2.x by creating my own formatter and passing it through.

cheetah-lint essentially takes a python-like templating language which compiles into python, compiles source files into python and then passes the python through flake8.  This makes it pretty easy to plug in existing python linting plugins targetting flake8 and utilize them to lint the templates.  You could imagine a similar setup to lint other template->python languages such as mako.

**Comment by asottile** [comment]
In GitLab by @sigmavirus24 on Dec 31, 2017, 16:59

https://github.com/asottile/yesqa/issues/1

**Comment by asottile** [comment]
In GitLab by @sigmavirus24 on Jan 1, 2018, 04:20

https://github.com/asottile/yesqa/issues/1#issuecomment-354631224

**Comment by asottile** [comment]
In GitLab by @martmists on Jan 20, 2018, 08:04

I think having a way to use Flake8 Similar to PyLint would help a lot already: `-f json` for JSON output, `pylint.lint.Run(..., exit=False)` can run pylint on files (though output is sent to stdout), but I'd say starting with simple things like this in the API would be a great start, and later on we could implement the more "complex" tasks. 

Running Flake8 on strings (or maybe even byte-like) would be a great improvement. Further improvements could be, as mentioned before, JSON output. To add to that, having a way to manually 'step' through lines in a file would give extensions and/or users access to more advanced ways of processing the data Flake8 returns.

**Comment by asottile** [comment]
In GitLab by @asottile on Oct 2, 2020, 16:38

changed the description

**Comment by asottile** [comment]
In GitLab by @asottile on Oct 21, 2020, 09:17

mentioned in merge request !458

**Comment by nbirillo** [comment]
Do you have any updates about this issue?

**Comment by asottile** [comment]
@nbirillo do you see any comments here?

**Comment by nbirillo** [comment]
@asottile No, that's why I'm asking. Maybe there is some news about the approximate milestone, whether the API will be developed at all, etc.

**Comment by asottile** [comment]
if there were news it would be in a comment here, obviously

**Comment by buhtz** [comment]
The only workaround I see currently is to run `flake8` via `subprocess.Popen()` and parse its output.
